lib/IRGen/GenCall.cpp

```diff
@@ -229,15 +229,16 @@ void irgen::addExtendAttribute(IRGenModule &IGM,
 namespace {
   class SignatureExpansion {
     IRGenModule &IGM;
+    GenericEnvironment *GenericEnv;
```

Comment:
Did you consider using the canonical generic environment associated with the module + generic signature, rather than propagating a generic environment throughout? Each `ArchetypeType` knows its generic environment (now), so one can map into that canonical generic environment.